universum/configuration_support.py

```diff
@@ -1,18 +1,210 @@
-from typing import Callable, Dict, Iterable, List, Union
+from typing import Any, Callable, Dict, Iterable, List, Optional, TypeVar, Union
 import copy
 import os
 
+
 __all__ = [
+    "ProjectConfiguration",
     "Variations",
-    "combine",
     "set_project_root",
     "get_project_root"
 ]
 
-skip_attributes = {"children", "critical", "skip_numbering_level"}
+
+class ProjectConfiguration:
+    """
+    ProjectConfiguration is a collection of configurable project-specific data needed for Universum operations.
+    Legacy constructor supports dictionary data, but for static type checking please use new explicit constructor.
+    """
+
+    # pylint: disable-msg=too-many-locals
+    def __init__(self,
+                 name: str = '',
+                 command: Optional[List[str]] = None,
+                 environment: Optional[Dict[str, str]] = None,
+                 artifacts: str = '',
+                 report_artifacts: str = '',
+                 artifact_prebuild_clean: bool = False,
+                 directory: str = '',
+                 critical: bool = False,
+                 background: bool = False,
+                 finish_background: bool = False,
+                 code_report: bool = False,
+                 pass_tag: str = '',
+                 fail_tag: str = '',
+                 if_env_set: str = '',
+                 extras: Optional[Dict[str, str]] = None,
+                 **kwargs) -> None:  # explicit constructor for client's type safety
+        self.name: str = name
+        self.directory: str = directory
+        self.code_report: bool = code_report
+        self.command: List[str] = command if command else []
+        self.environment: Dict[str, str] = environment if environment else {}
+        self.artifacts: str = artifacts
+        self.report_artifacts: str = report_artifacts
+        self.artifact_prebuild_clean: bool = artifact_prebuild_clean
+        self.critical: bool = critical
+        self.background: bool = background
+        self.finish_background: bool = finish_background
+        self.pass_tag: str = pass_tag
+        self.fail_tag: str = fail_tag
+        self.if_env_set: str = if_env_set
+        self.children: Optional['Variations'] = None
+        self.extras: Dict[str, str] = extras if extras else {}
+        for key, value in kwargs.items():
+            self.extras[key] = value
+
+    def __repr__(self) -> str:
```

Comment:
Why we need custom `__repr__` function at all?
Isn't the default implementation for class enough?

Reply:
default __repr__ returns "ProjectConfiguration object..."
Custom __repr__ helps with readable debug output + this way we can keep docstrings unchanged.

Reply:
Ok, I've got the debug output argument.
But what is the issue with docstrings exactly?

Reply:
in docstrings there are many cases where indexed Variations object is printed. Since indexing Variations should return ProjectConfiguration - we need to redefine repr so that comparisong with dict contents works in docstring tests

Reply:
Ok, as I understand, previously it printed an array of dicts with their exact content. This is what the default `__repr__` for dict does. And since we are printing list, its `__str__` method uses `__repr__` of its elements for printing, not `__str__`. Unfortunately, the default `__repr__` for class is useless.

Please expand the explanation in the docstring.

Reply:
exactly! I'll update the description